### DISCUSSION TITLE
Tasks can be mistakenly marked as "removed" in case of Deadlocks

### DISCUSSION THREAD

**Discussion by tanay27** [apache/airflow#32682] [General]
I am running Airflow 2.5.3 with Postgres as database.

In case of occasional Deadlocks that happen due to `UPDATE serialize_dag` query to store serialized dags and `SELECT...FOR UPDATE dag` query because of process based connection handling, the tasks are mistakenly marked as `removed` due to [this](https://github.com/apache/airflow/blob/2.6.3/airflow/models/dagrun.py#L719-L722) `try and except` block and mis-reports or does not attempt to run tasks that were queued.

I don't know exactly how to reproduce this since it happened with us on `production` and no way to tell exactly how to reproduce.

Good to know:

The DAG for which this behavior was observed was fairly large DAG with >90 task instances. 

What we did try as an immediate remedy is basically reducing scheduler count from 2 --> 1 and set `row_level_locking` to `false` but then there was another `SELECT...FOR UPDATE` query for a different table but we are not having dead runs since then, we still have ocassional deadlocks but scheduler is able to pick the task again.

I was thinking if we can somehow tackle this within code. 
In case of unsuccessful db query or disconnect, we should catch the error, maybe retry or do nothing and not mark the tasks as removed since they aren't technically removed.

What do you think?

**Comment by potiuk** [comment]
You can try to disable https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#schedule-after-task-execution. instead of decreasing scheduler count and see if it helps. It has been known to help in similar cases - and this is merely an optimization of task delays from the same DAGs. You can possobly try it and see if it helps and report back.
